p2p/src/connection/incoming/p2p_connection_incoming_reducer.rs

```diff
@@ -1,162 +1,533 @@
+use std::net::{IpAddr, SocketAddr};
+
+use multiaddr::Protocol;
+use openmina_core::{bug_condition, debug, warn, Substate};
+use redux::{ActionWithMeta, Dispatcher, Timestamp};
+
+use crate::{
+    connection::{
+        incoming::P2pConnectionIncomingError,
+        incoming_effectful::P2pConnectionIncomingEffectfulAction,
+        outgoing::P2pConnectionOutgoingInitOpts, P2pConnectionState,
+    },
+    disconnection::{P2pDisconnectionAction, P2pDisconnectionReason},
+    webrtc::{HttpSignalingInfo, SignalingMethod},
+    ConnectionAddr, P2pNetworkSchedulerAction, P2pPeerAction, P2pPeerState, P2pPeerStatus,
+    P2pState, PeerId,
+};
+
 use super::{
-    P2pConnectionIncomingAction, P2pConnectionIncomingActionWithMetaRef, P2pConnectionIncomingState,
+    super::{incoming::P2pConnectionIncomingState, RejectionReason},
+    IncomingSignalingMethod, P2pConnectionIncomingAction,
 };
 
 impl P2pConnectionIncomingState {
-    pub fn reducer(&mut self, action: P2pConnectionIncomingActionWithMetaRef<'_>) {
+    pub fn reducer<Action, State>(
+        mut state_context: Substate<Action, State, P2pState>,
```

Comment:
Please add a comment here explaining why this reducer is over `P2pState` instead of `Self`